balancer-js/src/modules/pools/pool-types/concerns/weighted/exit.concern.ts

```diff
@@ -0,0 +1,290 @@
+   * @param {Pool}      pool - Subgraph pool object of pool being exited
+   * @param {string[]}  tokensOut - Tokens provided for exiting pool
+   * @param {string[]}  amountsOut - Amoutns provided for exiting pool
+   * @param {string}    slippage - Maximum slippage tolerance in percentage. i.e. 0.05 = 5%
+   * @returns           transaction request ready to send with signer.sendTransaction
+   */
+  buildExitExactTokensOut = ({
+    exiter,
+    pool,
+    tokensOut,
+    amountsOut,
+    slippage,
+  }: ExitExactTokensOutParameters): ExitPoolAttributes => {
+    if (
+      tokensOut.length != amountsOut.length ||
+      tokensOut.length != pool.tokensList.length
+    ) {
+      throw new BalancerError(BalancerErrorCode.INPUT_LENGTH_MISMATCH);
+    }
+
+    const parsedPoolInfo = this.parsePoolInfo(pool);
+    const sortedPoolInfo = this.sortPoolInfo(
+      parsedPoolInfo.tokens,
+      parsedPoolInfo.balances,
+      parsedPoolInfo.weights,
+      parsedPoolInfo.decimals
+    );
+    const sortedInputs = this.sortInputs(tokensOut, amountsOut);
+
+    const bptIn = SDK.WeightedMath._calcBptInGivenExactTokensOut(
+      sortedPoolInfo.balances.map((b) => new OldBigNumber(b)),
+      sortedPoolInfo.weights.map((w) => new OldBigNumber(w)),
+      sortedInputs.amounts.map((a) => new OldBigNumber(a)),
+      new OldBigNumber(parsedPoolInfo.totalShares),
+      new OldBigNumber(parsedPoolInfo.swapFee)
+    ).toString();
+
+    const maxBPTIn = addSlippage(
+      BigNumber.from(bptIn),
+      BigNumber.from(slippage)
+    ).toString();
+
+    const userData = WeightedPoolEncoder.exitBPTInForExactTokensOut(
+      sortedInputs.amounts,
+      maxBPTIn
+    );
+
+    const to = balancerVault;
+    const functionName = 'exitPool';
+    const attributes: ExitPool = {
+      poolId: pool.id,
+      sender: exiter,
+      recipient: exiter,
+      exitPoolRequest: {
+        assets: sortedInputs.tokens,
+        minAmountsOut: sortedInputs.amounts,
+        userData,
+        toInternalBalance: false,
+      },
+    };
+    const vaultInterface = Vault__factory.createInterface();
+    // encode transaction data into an ABI byte string which can be sent to the network to be executed
+    const data = vaultInterface.encodeFunctionData(functionName, [
+      attributes.poolId,
+      attributes.sender,
+      attributes.recipient,
+      attributes.exitPoolRequest,
+    ]);
+
+    return {
+      to,
+      functionName,
+      attributes,
+      data,
+      minAmountsOut: sortedInputs.amounts,
+      maxBPTIn,
+    };
+  };
+
+  // Helper methods
+
+  /**
+   * Sort pool info alphabetically by token addresses as required by gerogeroman SDK
+   */
+  private sortPoolInfo = (
+    poolTokens: string[],
+    poolBalances: string[],
+    poolWeights: string[],
+    poolDecimals: number[]
+  ) => {
+    const WETH = '0x000000000000000000000000000000000000000F'; // TODO: check if it should be possible to exit with ETH instead of WETH
```

Comment:
@brunoguerios - it is possible to exit to ETH. Can you add this functionality along with some tests. Then this PR should be good to go. Thanks!

Reply:
For `exitExactBptIn`, since the user does not provide an array of assets to exit pool, I'll add a boolean input that indicates that intention. Something like `shouldUnwrapWethToEth` as I believe that `exitWithEth` may seem misleading compared to `singleTokenMaxOut` scenario. How does that sound @johngrantuk ?